adr/ADR-42.md

```diff
@@ -0,0 +1,168 @@
+
+## General Overview
+
+We introduce 2 settings on `ConsumerConfig` that activates these related features:
+
+```go
+{
+	PriorityGroups: ["<group>", ...],
+	PriorityPolicy: "<policy>"
+}
+```
+
+The presence of the `PriorityGroups` setting activates the set of features documented here with a list of names supported
+behaviours configured using the `PriorityPolicy`.
+
+Technically in some of the proposed policies the `PriorityGroups` have no real meaning today, but we keep it for consistency 
+and allow for future features to be added that would be per-group without then requiring all clients to be updated. 
+Future message grouping features would require groups to be listed here.
+
+In the initial implementation we should limit `PriorityGroups` to one per consumer only and error should one be made with 
+multiple groups. In future iterations multiple groups will be supported along with dynamic partitioning of stream
+data.
+
+This is only supported on Pull Consumers, configuring this on a Push consumer must raise an error.
+
+## Priority Policies
+
+### `overflow` policy
+
+Users want certain clients to only get messages when certain criteria are met.
+
+Imagine jobs are best processed locally in `us-east-1` but at times there might be so many jobs in that region that 
+`us-west-1` region handling some overflow, while less optimal in terms of transit costs and latency, would be desirable 
+to ensure serving client needs as soon as possible.
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+Here we state that the Consumer has one group called `jobs`, it is operating on the `overflow` policy and requires `explicit`
+Acks, any other Ack policy will produce an error. If we force this ack policy in normal use we should error in Pedantic mode.
+
+Pull requests will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"min_pending": 1000` - only deliver messages when `num_pending` for the consumer is >= 1000
+ * `"min_ack_pending: 1000` - only deliver messages when `ack_pending` for the consumer is >= 1000
+
+If `min_pending` and `min_ack_pending` are both given both have to be satisfied (boolean AND).
+
+### `pinned_client` policy
+
+Users want to have a single client perform all the work for a consumer, but they also want to have a stand-by client that
+can take over when the primary, aka `pinned` client, fails.
+
+Users would like the ability to influence which client will generally be considered the active, pinned, one - but we feel
```

Comment:
We agreed why this is the safe idea, and priority can have other uses too, however I wonder if two scenarios will pop up often:

**Users want to stepdown a pinned consumer administratively. Just like stream stepdown.**
With this design, every operation like this would mean lowering the priority. That could lead to priority drifing to 1 which would block the possibility to move further.

Maybe we can have this without any changes in ADR, just some new API call in the future?

Reply:
I mention here later a API to trigger a switch - but if people are fiddling priorities too much they'd just end up picking the one that just was the pinned one again. I am kind of less sold on the priority idea atm

Reply:
I might not add the priority idea in the initial build I think, what do you think @Jarema gives us time to think about this more. I'm definitely also worried about lots of unexpected switches of pinned client if someone just gets the timings a bit wrong and priorities will make that worse

Reply:
Yes, I agree. 
I really like that we're designing the features with future in mind, but we also do not have to implement those part of the design that are not required for first steps.

Addition of priority later should also not break anything, as we could say that not setting priority sets its to, for example, 5.

Reply:
@jnmoyne would you say being able to influence next-pinned-client by priority is important for initial work? We want some more time to explore how that will work

Reply:
No, for initial work I would want just pick any other available. People that are asking for priority would be satisfied with 'overflow'. IMHO trying to have both priority an exclusivity is an architectural mistake. If you do it, the next thing they will ask for is that if the pinned client has moved to a lower priority one because the higher priority one has left, then when one with a higher priority comes back up they will want it become again the new pinned client.
Otherwise you end up with the priority race condition you describe above. And if you do that then it is effectively 'overflow' all over again.

So to do 'exclusive with priority' you just use 'overflow' with just one instance per priority level (should be trivial to do with any container orchestration), therefore I don't think we would ever need to add priorities to pinned.